toolchain/parse/node_kind.cpp

```diff
@@ -38,12 +39,53 @@ auto NodeKind::bracket() const -> NodeKind {
 auto NodeKind::child_count() const -> int32_t {
   static constexpr int32_t ChildCount[] = {
 #define CARBON_PARSE_NODE_KIND_BRACKET(...) -1,
-#define CARBON_PARSE_NODE_KIND_CHILD_COUNT(Name, Size) Size,
+#define CARBON_PARSE_NODE_KIND_CHILD_COUNT(Name, Size, ...) Size,
 #include "toolchain/parse/node_kind.def"
   };
   auto child_count = ChildCount[AsInt()];
   CARBON_CHECK(child_count >= 0) << *this;
   return child_count;
 }
 
+void CheckNodeMatchesLexerToken(NodeKind node_kind, Lex::TokenKind token_kind,
+                                bool has_error) {
+  switch (node_kind) {
+#define CARBON_ANY_TOKEN return;
+
+#define CARBON_TOKEN(Expected)                  \
+  if (token_kind == Lex::TokenKind::Expected) { \
+    return;                                     \
+  }                                             \
+  break;
+
+#define CARBON_TOKEN_EITHER(Expected1, Expected2)               \
+  if (token_kind == Lex::TokenKind::Expected1 ||                \
+      (has_error && token_kind == Lex::TokenKind::Expected2)) { \
+    return;                                                     \
+  }                                                             \
+  break;
+
+#define CARBON_TOKEN_UNLESS_ERROR(Expected)                  \
+  if (has_error || token_kind == Lex::TokenKind::Expected) { \
+    return;                                                  \
+  }                                                          \
+  break;
+
+#define CARBON_CASE(Name, MatchAction) \
+  case NodeKind::Name:                 \
+    MatchAction
+
+#define CARBON_PARSE_NODE_KIND_BRACKET(Name, BracketName, MatchAction) \
+  CARBON_CASE(Name, MatchAction)
+
+#define CARBON_PARSE_NODE_KIND_CHILD_COUNT(Name, Size, MatchAction) \
+  CARBON_CASE(Name, MatchAction)
+
+#include "toolchain/parse/node_kind.def"
```

Comment:
I wonder if it'd be clearer to allow `MatchAction` to have a sequence of actions, and separate out the error handling into a separate action:
```suggestion
#define CARBON_ANY_TOKEN return;

#define CARBON_TOKEN(Expected)                  \
  if (token_kind == Lex::TokenKind::Expected) { \
    return;                                     \
  }

#define CARBON_IF_ERROR(MatchActions) \
  if (has_error) {                    \
    MatchActions                      \
  }

#define CARBON_CASE(Name, MatchActions) \
  case NodeKind::Name:                  \
    MatchActions                        \
    break;

#define CARBON_PARSE_NODE_KIND_BRACKET(Name, BracketName, MatchActions) \
  CARBON_CASE(Name, MatchActions)

#define CARBON_PARSE_NODE_KIND_CHILD_COUNT(Name, Size, MatchActions) \
  CARBON_CASE(Name, MatchActions)

#include "toolchain/parse/node_kind.def"
```

(Suggestion continued in `.def` file.)